### PR TITLE
Code generation and more flexible quantities

### PR DESCRIPTION
**Pull request by harrypuuter**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,10 +1,25 @@
...
-if (NOT DEFINED CONFIG)
-    message(FATAL_ERROR "Please specify the config to be used with -DCONFIG=my_config_name")
+if (NOT DEFINED ANALYSIS)
+    message(FATAL_ERROR "Please specify the config to be used with -DANALYSIS=my_config_name")
 endif()
-message(STATUS "Set up analysis for config ${CONFIG}.")
+message(STATUS "Set up analysis for config ${ANALYSIS}.")
+
+if (NOT DEFINED CHANNELS)
+    message("No channel specificed, build the four main channels using -DCHANNELS=et,mt,tt,em")
```
Comment (stwunsch) [inline CMakeLists.txt:10]: Muss hier `,` oder `;` hin?

File: code_generation/producer.py [hunk 2]
```diff
@@ -1,36 +1,83 @@
...
-        calls = [self.writecall()]
-        for shift in self.output.shifts:
-            calls.append(self.writecall(shift))
+        self.output.shift(
+            name
+        )  # crashes on purpose if output is None. This method should not be called for a producer without output
+
+    def writecall(self, config, shift=""):
+        config[shift]["output"] = (
+            "" if self.output == None else self.output.get_leaf(shift)
```
[the remark this answers is not in the document]
Comment (ArturAkh) [reply code_generation/producer.py]: Agreed. Committing that change to pull request

File: code_generation/producer.py [hunk 3]
```diff
@@ -1,36 +1,83 @@
...
+            '{"' + '","'.join([x.get_leaf(shift) for x in self.inputs]) + '"}'
+        )
+        config[shift]["df"] = "{df}"
+        return self.call.format(
+            **config[shift]
+        )  # use format (not format_map here) such that missing config entries cause an error
+
+    def writecalls(self, config):
+        calls = [self.writecall(config)]
+        if self.output != None:
```
[the remark this answers is not in the document]
Comment (ArturAkh) [reply code_generation/producer.py]: Agreed. Comitting.

File: code_generation/producer.py [hunk 4]
```diff
@@ -1,36 +1,83 @@
...
+            for shift in self.output.shifts:
+                calls.append(self.writecall(config, shift))
         return calls
 
-prod1=producer("phyticsd::FilterID(auto {df}, const std::string {output}, std::string isolationName)", [], q.pt_1)
-prod2=producer("FilterID(auto {df}, const std::string {output}, std::string isolationName)", [], q.pt_2)
-prod3=producer('FilterID({df}, "{output}", {input_coll}, {ptcut})', [q.pt_1, q.pt_2], q.m_vis)
+
+class producer_vec(producer):
```
[the remark this answers is not in the document]
Comment (ArturAkh) [reply code_generation/producer.py]: Agreed. Comitting

File: code_generation/producer.py [hunk 5]
```diff
@@ -1,36 +1,83 @@
...
+            shifts.extend(self.output.shifts)
+        for shift in shifts:
+            for val in config[shift][self.vec_config]:
+                self.call = basecall.format_map(SafeDict({self.vec_config: val}))
+                calls.append(self.writecall(config, shift))
+        self.call = basecall
+        return calls
+
+
+prod1 = producer(
```
[the remark this answers is not in the document]
Comment (ArturAkh) [reply code_generation/producer.py]: Agreed. Comitting

File: code_generation/producer.py [hunk 6]
```diff
@@ -1,36 +1,83 @@
...
+    "phyticsd::FilterID(auto {df}, const std::string {output}, std::string isolationName)",
+    [],
+    q.pt_1,
+)
+prod2 = producer(
```
[the remark this answers is not in the document]
Comment (ArturAkh) [reply code_generation/producer.py]: Agreed. Comitting

File: code_generation/producer.py [hunk 7]
```diff
@@ -1,36 +1,83 @@
...
+    "FilterID(auto {df}, const std::string {output}, std::string isolationName)",
+    [],
+    q.pt_2,
+)
+prod3 = producer(
```
[the remark this answers is not in the document]
Comment (ArturAkh) [reply code_generation/producer.py]: Agreed. Comitting

File: code_generation/producer.py [hunk 8]
```diff
@@ -1,36 +1,83 @@
...
+    'FilterID({df}, "{output}", {input_coll}, {ptcut})', [q.pt_1, q.pt_2], q.m_vis
+)
+
+MetFilter = producer_vec(
```
[the remark this answers is not in the document]
Comment (ArturAkh) [reply code_generation/producer.py]: Agreed. Comitting.

**Review by ArturAkh — APPROVED** [review]
I guess a lot of the changes are reformatting, right?

It is difficult to distinguish the real changes from reformatting, but I think the pull request is fine.

I've included suggested changes from @mburkart into the pull request

**Comment by harrypuuter** [comment]
Yes in hindsight, it was a dump idea to apply style changes in this PR as well ...
I am sorry

**Pull request merged**